src/main/scala/com.snowplowanalytics/iglu.schemaddl/migrations/SchemaList.scala

```diff
@@ -0,0 +1,286 @@
+    */
+  case class SchemaListSegment private(schemas: NonEmptyList[IgluSchema]) extends AnyVal
+  private[migrations] object SchemaListSegment
+
+  /**
+    * Represents errors while creating SchemaFullList
+    */
+  sealed trait FullListCreationError extends Product with Serializable
+
+  object FullListCreationError {
+
+    /**
+      * Returned when given model group have schemas which
+      * they could not be ordered unambiguously.
+      * For example, [1-0-0, 1-1-0, 1-0-2] schema list
+      * could not be ordered unambiguously because it could be
+      * either [1-0-0, 1-0-2, 1-1-0] or [1-0-0, 1-1-0, 1-0-2]
+      */
+    case class AmbiguousOrder(schemas: ModelGroupList) extends FullListCreationError
+
+    /**
+      * Returned when there is a gap in the schema list.
+      * [1-0-0, 1-0-1, 1-0-3] is missing 1-0-2 version,
+      * therefore gap error will be returned in this case
+      */
+    case class GapInModelGroup(schemas: ModelGroupList) extends FullListCreationError
+
+    /**
+      * Represents unexpected error cases while creating SchemaFullList
+      */
+    case class UnexpectedError(schemas: ModelGroupList) extends FullListCreationError
+  }
+
+  /**
+    * Always belong to the same vendor/name/model triple,
+    * however different from SchemaList, no ordering is implied.
+    */
+  sealed trait ModelGroupList {
+    val schemas: NonEmptyList[IgluSchema]
+
+    override def equals(that: Any): Boolean =
+      that match {
+        case that: ModelGroupList => {
+          that.schemas == schemas
+        }
+        case _ => false
+      }
+  }
+
+  object ModelGroupList {
+
+    /** Split schemas into a lists grouped by model group (still no order implied) */
+    def groupSchemas(schemas: NonEmptyList[IgluSchema]): NonEmptyList[ModelGroupList] =
+      schemas.groupByNem(schema => getModelGroup(schema.self)).toNel.map { case (_, nel) => new ModelGroupList { val schemas = nel } }
+  }
+
+  /**
+    * Fetch from Iglu Server and parse each schema from `SchemaKeyList`, using generic resolution function
+    * (IO-dependent) valid constructor of `SchemaList`
+    * @param keys non-empty properly ordered list of `SchemaKey`s, fetched from Iglu Server
+    * @param fetch resolution function
+    * @return properly ordered list of parsed JSON Schemas
+    */
+  def fromSchemaList[F[_]: Monad, E](keys: SchemaKeyList, fetch: SchemaKey => EitherT[F, E, IgluSchema]): EitherT[F, E, SchemaList] =
+    for {
+      schemas <- keys.schemas.traverse { key => fetch(key) }
+    } yield {
+      schemas match {
+        case Nil => throw new IllegalStateException("Result list can not be empty")
+        case h :: Nil => SingleSchema(h)
+        case h :: t => SchemaListFull(NonEmptyList(h, t))
+      }
+    }
+
+  /**
+    * Fetch NonEmptyList of IgluSchemas from Iglu Server, using generic resolution function
+    * @param fetch resolution function
```

Comment:
I think `fetch` in this case not a "resolution function":

1. Its a value (`IO`/`EitherT` object)
2. Resolution function resolves one schema - here we pull data from `/schemas` endpoint. We need to mention that.